### PR TITLE
Feature/mb add aa workflow

### PR DESCRIPTION
**Pull request by migbro**
## Description

This ties all the tools from the last pull request, plus a few updates into a single workflow. The readme has also been updated to explain inputs and outputs. After approval, final output and default inputs will be set for final release and inclusion into the monster somatic workflow, or a new subworkflow may be spawned as the main workflow already has a cram to bam step

Fixes # https://github.com/d3b-center/bixu-tracker/issues/1444

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes cwltool with acceptable warnings
- [ ] [T/N Cram input scenario](https://cavatica.sbgenomics.com/u/d3b-bixu/kf-amplicon-architect-dev/tasks/b0327d6e-8f9c-4ebc-b21e-fefc951bf31b/)
- [ ] [Existing .cns and bam input scenario](https://cavatica.sbgenomics.com/u/d3b-bixu/kf-amplicon-architect-dev/tasks/6f88fc8b-2641-4c6e-ab75-8ab0d7023b22/)
- [ ] [Expected mostly empty output scenario](https://cavatica.sbgenomics.com/u/d3b-bixu/kf-amplicon-architect-dev/tasks/40879ad7-1418-405a-a966-3c87e306d6b4/)
- [ ] [Repeat with unneeded input removed (ref_cache)](https://cavatica.sbgenomics.com/u/d3b-bixu/kf-amplicon-architect-dev/tasks/40879ad7-1418-405a-a966-3c87e306d6b4/stats/)

**Test Configuration**:
* Environment:
* Test files:

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings
- [ ] I have committed any related changes to the PR

### CODE REVIEW

File: tools/aa_classifier.cwl [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    coresMin: $(inputs.threads)
+  - class: InitialWorkDirRequirement
+    listing: [$(inputs.data_repo)]
+baseCommand: [export]
+arguments:
+  - position: 0
+    shellQuote: false
+    valueFrom: >-
+      AA_DATA_REPO=$(inputs.data_repo.path)
+      && export AA_SRC=/home/programs/AmpliconArchitect-master/src
```
Comment (dmiller15) [inline tools/aa_classifier.cwl:21]: this can be set with: https://www.commonwl.org/v1.2/CommandLineTool.html#EnvVarRequirement

File: docs/kfdrc_amplicon_architect.md [hunk 2]
```diff
@@ -4,6 +4,40 @@ Briefly, chromosomal fragments in tumor cells break off to form circular DNA.
...
+### Outputs:
+ - `aa_cnv_seeds`: Bed file with candidate regions to search
+ - `aa_summary`: Summary for all amplicons detected by AA
+ - `aa_cycles`: Text file for each amplicon listing the edges in the breakpoint graph, their categorization (sequence, discordant, concordant, source) and their copy counts
+ - `graph`: A text file for each amplicon listing the edges in the breakpoint graph, their categorization (sequence, discordant, concordant, source) and their copy counts
+ - `sv_pdf`: PDF image file displaying the SV view of AA
+ - `sv_png`: PNG image file displaying the SV view of AA
+ - `amplicon_classification_profiles`: Abstract classification of the amplicon
+ - `gene_list`: Genes present on amplicons with each classification
+ - `ecDNA_cts`: Not yet defined
```
Comment (sickler-alex) [inline docs/kfdrc_amplicon_architect.md:38]: Should this be updated or will ecDNA_cts be added in a new PR?

File: tools/amplicon_architect.cwl [hunk 3]
```diff
@@ -10,16 +10,14 @@ requirements:
   - class: ResourceRequirement
     ramMin: 16000
     coresMin: $(inputs.threads)
-baseCommand: [mkdir, data_repo]
+  - class: InitialWorkDirRequirement
+    listing: [$(inputs.data_repo)]
+baseCommand: [export]
```
Comment (dmiller15) [inline tools/amplicon_architect.cwl:15]: I'd almost rather have no baseCommand in this scenario. just toss the export into the argument.

File: tools/prepare_aa.cwl [hunk 4]
```diff
@@ -10,18 +10,36 @@ requirements:
...
     valueFrom: >-
-      && export AA_DATA_REPO=$PWD/data_repo
+      AA_DATA_REPO=$(inputs.data_repo.path)
       && export AA_SRC=/home/programs/AmpliconArchitect-master/src
       && touch $AA_DATA_REPO/coverage.stats
+      ${
+        if (inputs.mosek_license_file){
+        var cmd = "&& mkdir licenses && cp " + inputs.mosek_license_file.path + " licenses\
+        && export MOSEKLM_LICENSE_FILE=$PWD/licenses";
+        return cmd;
```
Comment (dmiller15) [inline tools/prepare_aa.cwl:28]: indendation is off.

File: tools/prepare_aa.cwl [hunk 5]
```diff
@@ -10,18 +10,36 @@ requirements:
...
+        }
+        else{
+          return ""
+        }
+      }
+      ${
+        if (inputs.run_ac){
+        var cmd = "&& export AC_SRC=/home/programs/AmpliconClassifier-main";
+        return cmd;
```
Comment (dmiller15) [inline tools/prepare_aa.cwl:37]: indentation

File: tools/prepare_aa.cwl [hunk 6]
```diff
@@ -32,18 +50,21 @@ arguments:
...
+  data_ref_version: { type: ['null', {type: enum, name: wgs_mode, symbols: ["GRCh38", "hg19", "GRCh37", "mm10", "GRCm38"]}], doc: "Genome version in data repo to use", default: "GRCh38", inputBinding: { position: 1, prefix: "--ref"} }
+  sorted_bam: { type: File, doc: "tumor bam file", secondaryFiles: ['^.bai?', '.crai?'], inputBinding: { position: 1, prefix: "--sorted_bam" } }
+  sample: { type: 'string', doc: "Sample name", inputBinding: { position: 1, prefix: "-s"} }
+  threads: { type: 'int?', doc: 'Num threads to use', default: 8, inputBinding: { position: 1, prefix: "-t"} }
+  cnv_bed: { type: File, doc: "Converted CNVkit cns-to-bed file", inputBinding: { position: 1, prefix: "--cnv_bed"} }
+  run_aa: { type: 'boolean?', doc: "Flag to just go ahead and run amplicon amplifier right after", default: false, inputBinding: { position: 1, prefix: "--run_AA"} }
+  run_ac: { type: 'boolean?', doc: "Flag to just go ahead and run amplicon classifier right after", default: false, inputBinding: { position: 1, prefix: "--run_AC"} }
```
Comment (dmiller15) [inline tools/prepare_aa.cwl:66]: I think we talked about this but default false with an optional boolean is redundant. doesn't hurt though.

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+requirements:
+- class: MultipleInputFeatureRequirement
+- class: InlineJavascriptRequirement
+- class: ScatterFeatureRequirement
+
+inputs:
+  # Required
+  aa_data_repo: {type: File, doc: "Reference tar ball obtained from https://datasets.genepattern.org/?prefix=data/module_support_files/AmpliconArchitect/"}
+  aa_data_ref_version: {type: ['null', {type: enum, name: wgs_mode, symbols: ["GRCh38",
```
Comment (dmiller15) [inline workflow/kfdrc_production_amplicon_architect.cwl:147]: why is it named wgs_mode?

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+  amplicon_classification_profiles: {type: 'File[]?', doc: "abstract classification\
+      \ of the amplicon", outputSource: amplicon_classifier/amplicon_classification_profiles}
+  gene_list: {type: 'File[]?', doc: "genes present on amplicons with each classification",
+    outputSource: amplicon_classifier/gene_list}
+  ecDNA_cts: {type: 'File[]?', doc: "not yet defined", outputSource: amplicon_classifier/ecDNA_cts}
+
+steps:
+  samtools_tumor_cram_to_bam:
+    run: ../tools/samtools_calmd.cwl
+    when: $(inputs.input_reads.nameext == ".cram")
```
Comment (sickler-alex) [inline workflow/kfdrc_production_amplicon_architect.cwl:191]: Should this also check that input_reads != null?

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+      \ not already built", secondaryFiles: [.fai]}
+  # If proper .cns file exists
+  cnvkit_cns: {type: 'File?', doc: ".cns file from previous CNVkit run, if available.\
+      \ DO NOT USE .call.cns"}
+  # If CNVkit needs running and .cnn file exists
+  cnvkit_cnn: {type: 'File?', doc: "If running using an existing .cnn, supply here"}
+  # If CNVkit needs running and .cnn NOT available
+  normal_align_file: {type: 'File?', doc: "Normal read alignment file. Can cram or\
+      \ bam", secondaryFiles: [{pattern: '^.bai', required: false}, {pattern: '.crai',
+        required: false}]}
```
Comment (dmiller15) [inline workflow/kfdrc_production_amplicon_architect.cwl:166]: you're bouncing between notations here. i can tolerate one or the other but not both.

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+  sv_png: {type: 'File[]', doc: "PNG image file displaying the SV view of AA", outputSource: amplicon_architect/sv_png}
+  amplicon_classification_profiles: {type: 'File[]?', doc: "abstract classification\
+      \ of the amplicon", outputSource: amplicon_classifier/amplicon_classification_profiles}
+  gene_list: {type: 'File[]?', doc: "genes present on amplicons with each classification",
+    outputSource: amplicon_classifier/gene_list}
+  ecDNA_cts: {type: 'File[]?', doc: "not yet defined", outputSource: amplicon_classifier/ecDNA_cts}
+
+steps:
+  samtools_tumor_cram_to_bam:
+    run: ../tools/samtools_calmd.cwl
```
Comment (dmiller15) [inline workflow/kfdrc_production_amplicon_architect.cwl:190]: Maybe rename the step to note that you're actually recalculating the MD values.

**Review by sickler-alex — COMMENTED** [review]
One suggestion would be to make cnvkit a subworkflow and check for the cnn file when calling the subworkflow so you can skip the samtools steps as well as the cnvkit step.

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+
+  untar_data_repo:
+    run: ../tools/untar_gzip.cwl
+    in:
+      input_tar: aa_data_repo
+      out_dir_name:
+        valueFrom: ${return "data_repo"}
+    out: [output]
+
+
```
Comment (dmiller15) [inline workflow/kfdrc_production_amplicon_architect.cwl:213]: bonus empty line

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+        source: [samtools_tumor_cram_to_bam/bam_file, tumor_align_file]
+        pickValue: first_non_null
+      input_control:
+        source: [samtools_normal_cram_to_bam/bam_file, normal_align_file]
+        pickValue: first_non_null
+      reference:
+        # samtools needs fasta, can't use fasta if cnn file provided
+        source: [cnvkit_cnn, reference]
+        valueFrom: "${ if (self[0] == null){ return self[1]; } else{ return null;\
+          \ } }"
```
Comment (dmiller15) [inline workflow/kfdrc_production_amplicon_architect.cwl:228]: I'm pretty sure you can use the `inputs` object here. According to the [spec](https://www.commonwl.org/v1.2/Workflow.html#WorkflowStepInput):
> The value of `inputs` in the parameter reference or expression must be the input object to the workflow step after assigning the `source` values, applying `default`, and then scattering.

Since you are passing in `cnvkit_cnn` below you can set this to:  
```
source: reference
valueFrom: $(inputs.cnvkit_cnn == null ? self : null)
```
Also, the ternary operator is your friend: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+      graph: amplicon_architect/graph
+    scatter: [cycles, graph]
+    scatterMethod: dotproduct
+    out: [amplicon_classification_profiles, gene_list, ecDNA_cts]
+
+$namespaces:
+  sbg: https://sevenbridges.com
+hints:
+- class: "sbg:maxNumberOfParallelInstances"
+  value: 3
```
Comment (dmiller15) [inline workflow/kfdrc_production_amplicon_architect.cwl:284]: Looking at your test runs you never use more than 1 instance at a time. Do you really need three parallel for this?

File: workflow/kfdrc_production_amplicon_architect.cwl [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+"sbg:license": Apache License 2.0
+"sbg:publisher": KFDRC
```
Comment (dmiller15) [inline workflow/kfdrc_production_amplicon_architect.cwl:286]: I feel like you should add some tags to this.

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 8 above)
Comment (dmiller15) [reply workflow/kfdrc_production_amplicon_architect.cwl:191]: It's a required input. It's checked on input.

**Review by dmiller15 — CHANGES_REQUESTED** [review]

**Comment by dmiller15** [comment]
Another thing was resource declarations. Not really necessary to change now but something to keep in mind for future modification. You sound pretty confident that we won't need to modify them so it can pass for now.

File: docs/kfdrc_amplicon_architect.md (hunk 2 above)
Comment (migbro) [reply docs/kfdrc_amplicon_architect.md:38]: I will address in a new PR once our collaborators look it over

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 8 above)
Comment (migbro) [reply workflow/kfdrc_production_amplicon_architect.cwl:191]: Yup, that

File: tools/aa_classifier.cwl (hunk 1 above)
Comment (migbro) [reply tools/aa_classifier.cwl:21]: I was thinking about that ,but then got too deep. Def would be cleaner, so I will switch and hope this isn't oddly shell-dependent

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 7 above)
Comment (migbro) [reply workflow/kfdrc_production_amplicon_architect.cwl:147]: Ah, that's a holdover from copying from the CNVkit workflow

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 9 above)
Comment (migbro) [reply workflow/kfdrc_production_amplicon_architect.cwl:166]: Yes chef!

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 13 above)
Comment (migbro) [reply workflow/kfdrc_production_amplicon_architect.cwl:284]: I think in one of the tests I do use at least 2 instances - to convert the tumor and normal cram to bam at the same time

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 14 above)
Comment (migbro) [reply workflow/kfdrc_production_amplicon_architect.cwl:286]: I might if this goes as it's own public release, aside form being incorporated into the main workflow

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 8 above)
Comment (sickler-alex) [reply workflow/kfdrc_production_amplicon_architect.cwl:191]: Ok, then it should be removed in the step after this one.

File: tools/prepare_aa.cwl [hunk 15]
```diff
@@ -10,18 +10,36 @@ requirements:
...
+        else{
+          return ""
+        }
+      }
+      ${
+        if (inputs.run_ac){
+        var cmd = "&& export AC_SRC=/home/programs/AmpliconClassifier-main";
+        return cmd;
+        }
+        else{
```
Comment (luederm) [inline tools/prepare_aa.cwl:39]: Initially I was finding some of the more complex valueFroms hard to follow. I wonder if creating a bash wrapper at runtime could be more straightforward, as shown here: https://www.commonwl.org/user_guide/14-runtime/index.html

**Review by luederm — APPROVED** [review]
Looks good. I like the detailed documentation.

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 8 above)
Comment (dmiller15) [reply workflow/kfdrc_production_amplicon_architect.cwl:191]: No it shouldn't the normal_align_file is optional. You need to check if it's there.

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 14 above)
Comment (dmiller15) [reply workflow/kfdrc_production_amplicon_architect.cwl:286]: Ok. That's fine then. Seemed like you were trying to make this workflow a standalone. If not, then :+1:

File: workflow/kfdrc_production_amplicon_architect.cwl (hunk 13 above)
Comment (dmiller15) [reply workflow/kfdrc_production_amplicon_architect.cwl:284]: Sounds good.

File: tools/prepare_aa.cwl (hunk 15 above)
Comment (dmiller15) [reply tools/prepare_aa.cwl:39]: A good suggestion. Especially for complex steps that use pipefail and a lot of different tools.

Comment (migbro) [reply tools/prepare_aa.cwl:32]: Interesting thought, so, more like, when I have these java script conditionals, create a shell script at the beginning that will be run if the conditional is met? Like for instance: 
```yaml
      ${
        if (inputs.mosek_license_file){
          var cmd = "&& mkdir licenses && cp " + inputs.mosek_license_file.path + " licenses\
          && export MOSEKLM_LICENSE_FILE=$PWD/licenses";
          return cmd;
        }
        else{
          return ""
        }
      }
```

becomes:
```yaml
    listing:
      - entryname: cp_license.sh
        entry: |-
          mkdir license
          cp $(inputs.mosek_license_file.path licenses\
          export MOSEKLM_LICENSE_FILE=\$PWD/licenses
```
as the script, and the js conditional becomes:
```js
      ${
        if (inputs.mosek_license_file){
          return " && bash cp_license.sh";
      }
     else{
      return "";
     }
```
Something like that? Or also replace all the conditionals/section with a big shell script?

Comment (luederm) [reply tools/prepare_aa.cwl:32]: My thought was the latter, to include the conditionals in the shell script

Comment (migbro) [reply tools/prepare_aa.cwl]: Ok, it  was a little tricky, but I updated the three tools that had exported and replaced them with shell s scripts that get sourced. Downside is if the script fails, not sure how well the error is caught

**Comment by migbro** [comment]
@dmiller15 , I think I addressed your comments. Also testing a re-run here:
https://cavatica.sbgenomics.com/u/d3b-bixu/kf-amplicon-architect-dev/tasks/00ccf33f-9543-4583-9d16-bc1d8f755216/
and 
https://cavatica.sbgenomics.com/u/d3b-bixu/kf-amplicon-architect-dev/tasks/cecf5cb5-0961-4181-8249-8f0e5e952a1b/

**Review by dmiller15 — APPROVED** [review]

**Review by sickler-alex — APPROVED** [review]
